peer.go

```diff
@@ -1,516 +1,518 @@
+		tx := btcutil.NewTx(msg)
+		iv := wire.NewInvVect(wire.InvTypeTx, tx.Sha())
+		p.AddKnownInventory(iv)
+
+		// Queue the transaction up to be handled by the block manager and
+		// intentionally block further receives until the transaction is fully
+		// processed and known good or bad.  This helps prevent a malicious peer
+		// from queueing up a bunch of bad transactions before disconnecting (or
+		// being disconnected) and wasting memory.
+		s.blockManager.QueueTx(tx, p)
+		p.WaitForTx()
+	}
+}
+
+// handleBlockMsg is invoked when a peer receives a block bitcoin message.
+// It blocks until the bitcoin block has been fully processed.
+func handleBlockMsg(s *server) func(p *peer.Peer, msg *wire.MsgBlock, buf []byte) {
+	return func(p *peer.Peer, msg *wire.MsgBlock, buf []byte) {
+		// Convert the raw MsgBlock to a btcutil.Block which provides some
+		// convenience methods and things such as hash caching.
+		block := btcutil.NewBlockFromBlockAndBytes(msg, buf)
+
+		// Add the block to the known inventory for the peer.
+		iv := wire.NewInvVect(wire.InvTypeBlock, block.Sha())
+		p.AddKnownInventory(iv)
+
+		// Queue the block up to be handled by the block
+		// manager and intentionally block further receives
+		// until the bitcoin block is fully processed and known
+		// good or bad.  This helps prevent a malicious peer
+		// from queueing up a bunch of bad blocks before
+		// disconnecting (or being disconnected) and wasting
+		// memory.  Additionally, this behavior is depended on
+		// by at least the block acceptance test tool as the
+		// reference implementation processes blocks in the same
+		// thread and therefore blocks further messages until
+		// the bitcoin block has been fully processed.
+		s.blockManager.QueueBlock(block, p)
+		p.WaitForBlock()
+	}
+}
+
+// handleInvMsg is invoked when a peer receives an inv bitcoin message and
+// is used to examine the inventory being advertised by the remote peer and
+// react accordingly.  We pass the message down to blockmanager which will
+// call QueueMessage with any appropriate responses.
+func handleInvMsg(s *server) func(p *peer.Peer, msg *wire.MsgInv) {
+	return func(p *peer.Peer, msg *wire.MsgInv) {
+		s.blockManager.QueueInv(msg, p)
+	}
+}
+
+// handleHeadersMsg is invoked when a peer receives a headers bitcoin
+// message.  The message is passed down to the block manager.
+func handleHeadersMsg(s *server) func(p *peer.Peer, msg *wire.MsgHeaders) {
+	return func(p *peer.Peer, msg *wire.MsgHeaders) {
+		s.blockManager.QueueHeaders(msg, p)
+	}
+}
+
+// handleGetData is invoked when a peer receives a getdata bitcoin message
+// and is used to deliver block and transaction information.
+func handleGetDataMsg(s *server) func(p *peer.Peer, msg *wire.MsgGetData) {
+	return func(p *peer.Peer, msg *wire.MsgGetData) {
+		numAdded := 0
+		notFound := wire.NewMsgNotFound()
+
+		// We wait on the this wait channel periodically to prevent queueing
```

Comment:
I realize this was already there, but might as well fix the comment while here.  `the this` -> `this`